pkg/reconciler/delivery/delivery.go

```diff
@@ -0,0 +1,139 @@
+	ReconcilerName = "Delivery"
+	// KCDNamespace is the namespace of this project (KCD = Knative Continuous Delivery)
+	KCDNamespace = "knative-serving"
+	// KCDName is the name of this project
+	KCDName = "knative-continuous-delivery"
+)
+
+// Reconciler implements controller.Reconciler
+type Reconciler struct {
+	client clientset.Interface
+
+	configurationLister listers.ConfigurationLister
+	revisionLister      listers.RevisionLister 
+	routeLister         listers.RouteLister 
+}
+
+// Check that our Reconciler implements ksvcreconciler.Interface
+var _ configurationreconciler.Interface = (*Reconciler)(nil)
+var (
+	newPercent int64 = 50
+	oldPercent int64 = 50
+)
+
+// ReconcileKind is a very simple proof-of-concept reconciliation method
+// Assumes that there is one existing Revision and one new Revision
+// when the new Revision arrives, split traffic between 2 Revisions
+// according to newPercent/oldPercent
+func (c *Reconciler) ReconcileKind(ctx context.Context, cfg *v1.Configuration) pkgreconciler.Event {
+	logger := logging.FromContext(ctx)
+
+	// ignore changes triggered by continuous-delivery itself
+	if cfg.Namespace == KCDNamespace && cfg.Name == KCDName {
+		return nil
+	}
+
+	// retrieve information about latest ready and created Revisions
+	// do nothing if latest created Revision is not ready
+	latestReady := cfg.Status.LatestReadyRevisionName
+	latestCreated := cfg.Status.LatestCreatedRevisionName
+	if latestReady != latestCreated {
+		return nil
+	}
+
+	// find the Route object
+	r, err := c.routeLister.Routes(cfg.Namespace).Get(cfg.Name)
+	if err != nil {
+		return err
+	}
+
+	// determine if the current routing status includes the latest Revision
+	// if yes, do nothing; if not, rewrite routing spec
+	route := r.DeepCopy()
+	for idx := range route.Status.Traffic {
+		if route.Status.Traffic[idx].RevisionName == latestReady {
+			return nil
+		}
+	}
+	if len(route.Status.Traffic) >= 2 {
+		logger.Info("Unsupported use case: current implementation only supports 2 Revisions at once")
+		return nil
+	}
+	route = modifyRouteSpec(route, latestReady)
+
+	// push the changed Route from Go memory to K8s
+	route, err = c.client.ServingV1().Routes(cfg.Namespace).Update(route)
+	return err
+}
+
+// modifyRouteSpec is a toy function that is designed specifically for the proof-of-concept
+// it modifies the Route spec field to accommodate the new Revision, if necessary
+func modifyRouteSpec(route *v1.Route, newRevName string) *v1.Route {
+	// if there is currently zero traffic targets, then set the Configuration's
+	// latest ready Revision as the default traffic target
+	// if there is currently one traffic target, then split 50% off that target and
+	// direct it to the new Revision
+	if len(route.Status.Traffic) == 0 {
+		route.Spec.Traffic = []v1.TrafficTarget{
+			{
+				ConfigurationName: route.Name, // assume namespace/name matches for Route & Config
+				LatestRevision: boolPtr(true),
+				Percent: int64Ptr(100),
```

Comment:
This exists in knative pkg: https://github.com/knative/serving/blob/2c32e738d1bf1df4ca619f6d331777caca35f4d7/pkg/apis/serving/v1/route_validation_test.go#L51

Reply:
Nice library! I deleted my own versions of the ptr functions and used the ptr package instead.